backend-api/src/functions/adapters/dynamoDbAdapter.ts

```diff
@@ -0,0 +1,118 @@
+
+export class DynamoDbAdapter {
+  private readonly tableName: string;
+  private readonly dynamoDbClient = new DynamoDBClient({
+    region: process.env.REGION,
+    maxAttempts: 2,
+    requestHandler: new NodeHttpHandler({
+      connectionTimeout: 5000,
+      requestTimeout: 5000,
+    }),
+  });
+
+  constructor(tableName: string) {
+    this.tableName = tableName;
+  }
+
+  async getActiveSession(
+    subjectIdentifier: string,
+    attributesToGet: string[],
+  ): Promise<DatabaseRecord | null> {
+    const input: QueryCommandInput = {
+      TableName: this.tableName,
+      IndexName: "subjectIdentifier-timeToLive-index",
+      KeyConditionExpression:
+        "subjectIdentifier = :subjectIdentifier and :currentTimeInSeconds < timeToLive",
+      FilterExpression: "sessionState = :sessionState",
+      ExpressionAttributeValues: {
+        ":subjectIdentifier": { S: subjectIdentifier },
+        ":sessionState": { S: sessionStates.ASYNC_AUTH_SESSION_CREATED },
+        ":currentTimeInSeconds": {
+          N: this.getTimeNowInSeconds().toString(),
+        },
+      },
+      ProjectionExpression: this.formatAttributesToGet(attributesToGet),
+      Limit: 1,
+      ScanIndexForward: false,
+    };
+
+    const queryCommandOutput: QueryCommandOutput =
+      await this.dynamoDbClient.send(new QueryCommand(input));
+
+    const items = queryCommandOutput.Items;
+    if (!items || items.length === 0) {
+      return null;
+    }
+
+    return unmarshall(items[0]);
+  }
+
+  async createSession(
+    attributes: CreateSessionAttributes,
+    sessionId: string,
+  ): Promise<void> {
+    const {
+      client_id,
+      govuk_signin_journey_id,
+      issuer,
+      redirect_uri,
+      sessionDurationInSeconds,
+      state,
+      sub,
+    } = attributes;
+    const timeToLive = this.getTimeNowInSeconds() + sessionDurationInSeconds;
+
+    const input: PutItemCommandInput = {
+      TableName: this.tableName,
+      Item: {
+        clientId: { S: client_id },
+        govukSigninJourneyId: { S: govuk_signin_journey_id },
+        createdAt: { N: Date.now().toString() },
+        issuer: { S: issuer },
+        sessionId: { S: sessionId },
+        sessionState: { S: sessionStates.ASYNC_AUTH_SESSION_CREATED },
+        state: { S: state },
+        subjectIdentifier: { S: sub },
+        timeToLive: { N: timeToLive.toString() },
+        ...(redirect_uri && { redirectUri: { S: redirect_uri } }),
+      },
+      ConditionExpression: "attribute_not_exists(sessionId)",
+    };
+
+    try {
+      await this.dynamoDbClient.send(new PutItemCommand(input));
+    } catch (error) {
+      if (error instanceof ConditionalCheckFailedException) {
+        throw new Error("Session already exists with this ID");
+      } else {
+        throw error;
+      }
+    }
+  }
+
+  private getTimeNowInSeconds() {
+    return Math.floor(Date.now() / 1000);
+  }
+
+  formatAttributesToGet(attributes: string[]): string {
```

Comment:
please may we move this up to where it's used so it's a bit easier to look at the usage/implementation at the same time

Reply:
could this be a private function?

Reply:
As soon as I looked at this function alongside where it's used, the naming became crystal clear

Reply:
I wanted to have all helpers/private methods at the bottom 😞  - you will need to "find" this method within the file anyway

Reply:
I thought it was private!!!! Sorry :(

Reply:
Maybe the name should be formatAsProjectionExpression() ... @sandymay-dd ?

Reply:
I think that's much better...